util/log/structured.go

```diff
@@ -0,0 +1,168 @@
+// Copyright 2015 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Spencer Kimball (spencer.kimball@gmail.com)
+
+package log
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"reflect"
+	"unicode/utf8"
+
+	"github.com/cockroachdb/cockroach/proto"
+	gogoproto "github.com/gogo/protobuf/proto"
+	"golang.org/x/net/context"
+)
+
+// AddStructured creates a structured log entry to be written to the
+// specified facility of the logger.
+func AddStructured(ctx context.Context, s severity, depth int, format string, args []interface{}) {
+	file, line := Caller(depth + 1)
+	entry := &proto.LogEntry{}
+	setLogEntry(ctx, format, args, entry)
+	logging.outputLogEntry(s, file, line, false, entry)
+}
+
+// getJSON returns a JSON representation of the specified argument.
+// Returns nil if the type is simple and does not require a separate
+// JSON representation.
+func getJSON(arg interface{}) []byte {
+	// Handle simple types.
+	switch arg.(type) {
+	case bool, float32, float64, int, int8, int16, int32, int64, uint, uint8, uint16, uint32, uint64, uintptr, string, []byte:
+		return nil
+	}
+
+	jsonBytes, err := json.Marshal(arg)
+	if err != nil {
+		return []byte(fmt.Sprintf("{\"error\": %q}", err.Error()))
+	}
+	return jsonBytes
+}
+
+// setLogEntry populates the log entry with format, arguments,
+// and any available fields set in the context.
+func setLogEntry(ctx context.Context, format string, args []interface{}, entry *proto.LogEntry) {
+	entry.Format, entry.Args = parseFormatWithArgs(format, args)
+
+	if ctx != nil {
+		for i := Field(0); i < maxField; i++ {
+			if v := ctx.Value(i); v != nil {
+				switch i {
+				case NodeID:
+					entry.NodeID = gogoproto.Int32(int32(v.(proto.NodeID)))
+				case StoreID:
+					entry.StoreID = gogoproto.Int32(int32(v.(proto.StoreID)))
+				case RaftID:
+					entry.RaftID = gogoproto.Int64(v.(int64))
+				case Method:
+					entry.Method = gogoproto.Int32(int32(v.(proto.Method)))
+				case Key:
+					entry.Key = v.(proto.Key)
+				}
+			}
+		}
+	}
+}
+
+// parseFormatWithArgs parses the format string, matching each
+// format specifier with an argument from the args array.
+func parseFormatWithArgs(format string, args []interface{}) (string, []proto.LogEntry_Arg) {
```

Comment:
By the way, this code is taken from go's fmt/print.go and I don't see any reason to expect that standard to evolve meaningfully.